Multimodal integration - pixtral/llava/qwen2-vl

bursteratom opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description

rudimentary support for pixtral/llava/qwen2-vl

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

## Social Handles (Optional)

<!-- Thanks for submitting a bugfix or enhancement. -->
<!-- We'd love to show our thanks to you on Twitter & Discord if you provide your handle -->

bursteratom committed:
multimodal example config

bursteratom committed:
rebased multimodal support - lint

bursteratom committed:
more rebasing

bursteratom committed:
Merge branch 'main' into pixtral_integration2

bursteratom committed:
Merge branch 'main' into pixtral_integration2

NanoCode012 commented on src/axolotl/utils/collators/mm_chat.py:
```diff
@@ -35,142 +36,187 @@ def torch_call(
+            # Create new dict without 'conversations' key
+            result = deepcopy(example)
+            result.pop("conversations")
+            return {"messages": messages, **result}
```
```suggestion
            result.pop("conversations")
            result["messages"] = messages
            return result
```

There is an edge case here which I was told about recently. There could be a chance where `result` has `messages=None` which would overwrite the new `messages` key.

bursteratom committed:
Update src/axolotl/utils/collators/mm_chat.py

takes care of scenarios where result has messages=None which would overwrite the new messages key

Co-authored-by: NanoCode012 <nano@axolotl.ai>

bursteratom committed:
abstraction WIP

bursteratom committed:
higher level abstraction wip

bursteratom committed:
higher level abstraction wip

bursteratom committed:
higher level abstraction wip

bursteratom committed:
qwen2vl get text config

bursteratom committed:
qwen2vl get text config

bursteratom committed:
lint

bursteratom committed:
added docstring

winglian commented on src/axolotl/utils/collators/mm_processing_strategies.py:
Not sure this belongs in the collators module

bursteratom committed:
moved processing_strategies.py location

bursteratom committed:
Merge branch 'main' into pixtral_integration2

bursteratom committed:
Merge branch 'main' into pixtral_integration2

djsaunde commented on src/axolotl/processing_strategies/__init__.py:
```diff
@@ -0,0 +1,210 @@
+        return texts
+
+
+class Qwen2VLProcessingStrategy(ProcessingStrategy):
```
Seems like this isn't used anywhere?

bursteratom replied:
Thanks for catching that, added it to `get_processing_strategy`

djsaunde commented on src/axolotl/processing_strategies/__init__.py:
```diff
@@ -0,0 +1,210 @@
+        Args:
+        examples: list of conversation dictionaries
+        Returns:
+            dict in OpenAI format with 'messages' key
```
nit:

```suggestion
        Args:
            examples: list of conversation dictionaries
        
        Returns:
            dict in OpenAI format with 'messages' key
```

djsaunde commented on src/axolotl/processing_strategies/__init__.py:
```diff
@@ -0,0 +1,210 @@
+                self.image_token
+            )
+        except AttributeError:
+            pass
```
Maybe:

```suggestion
        self.image_token = None
        self.image_token_id = None
        
        if hasattr(processor, "image_token"):
            self.image_token = processor.image_token
            self.image_token_id = processor.tokenizer.convert_tokens_to_ids(
                self.image_token
            )
```

djsaunde commented on src/axolotl/processing_strategies/__init__.py:
```diff
@@ -0,0 +1,210 @@
+            else:
+                raise ValueError(
+                    "Only `messages` and `conversations` message keys are currently supported."
+                )
```
nit:

```suggestion
            if "messages" in example:  # OpenAI format
                processed_examples.append(example)
            elif "conversations" in example:  # Legacy format
                processed_examples.append(convert_legacy_format(example))
            else:
                raise ValueError(
                    "Only `messages` and `conversations` message keys are currently supported."
                )
```

djsaunde commented on src/axolotl/utils/collators/mm_chat.py:
```diff
@@ -35,154 +34,42 @@ def torch_call(
+        processing_strategy: ProcessingStrategy,
+        max_images,
+        length_only=False,
+    ):
```
nit: add type hints

bursteratom replied:
added. Thanks!

djsaunde commented on src/axolotl/utils/models.py:
```diff
@@ -367,7 +370,11 @@ def __init__(
+                self.text_model_config = self.model_config.text_config
+            except AttributeError:
+                # for qwen2_vl
+                self.text_model_config = self.model_config.get_text_config()
```
```suggestion
            if hasattr(self.model_config, "text_config"):
                self.text_model_config = self.model_config.text_config
            else:
                # for qwen2_vl
                self.text_model_config = self.model_config.get_text_config()
```

djsaunde commented on src/axolotl/utils/collators/mm_chat.py:
```diff
@@ -35,154 +34,42 @@ def torch_call(
         # The labels are the input_ids, and we mask the padding tokens in the loss computation
         labels = batch["input_ids"].clone()
-        labels[labels == processor.tokenizer.pad_token_id] = -100  #
+        labels[labels == processing_strategy.processor.tokenizer.pad_token_id] = -100  #
```
```suggestion
        labels[labels == processing_strategy.processor.tokenizer.pad_token_id] = -100
```

djsaunde reviewed (approved):
Nicely done, mostly just added some nits.

bursteratom committed:
Update src/axolotl/processing_strategies/__init__.py

Co-authored-by: Dan Saunders <danjsaund@gmail.com>

bursteratom committed:
Update src/axolotl/processing_strategies/__init__.py

Co-authored-by: Dan Saunders <danjsaund@gmail.com>

bursteratom committed:
Update src/axolotl/utils/models.py

Co-authored-by: Dan Saunders <danjsaund@gmail.com>

bursteratom committed:
Update src/axolotl/utils/collators/mm_chat.py

Co-authored-by: Dan Saunders <danjsaund@gmail.com>

bursteratom committed:
Update src/axolotl/processing_strategies/__init__.py

Co-authored-by: Dan Saunders <danjsaund@gmail.com>

bursteratom committed:
return Qwen2VLProcessingStrategy

bursteratom committed:
add type hint

bursteratom committed:
Merge branch 'main' into pixtral_integration2

bursteratom committed:
Merge branch 'main' into pixtral_integration2

bursteratom committed:
Merge branch 'main' into pixtral_integration2

bursteratom committed:
qwen2vl conditional generation

bursteratom committed:
Merge branch 'main' into pixtral_integration2

NanoCode012 committed:
feat: add qwen2_5_vl to multimodal list

bursteratom committed:
Merge branch 'main' into pixtral_integration2

bursteratom committed:
Merge branch 'main' into pixtral_integration2

ehartford commented:
@NanoCode012 @winglian is this waiting for something?

because, it seems like right now, llama-factory or just custom training code are the only way to train qwen2.5-vl

bursteratom committed:
Merge branch 'main' into pixtral_integration2

bursteratom committed:
Merge branch 'main' into pixtral_integration2

NanoCode012 commented on src/axolotl/utils/config/__init__.py:
```diff
@@ -170,7 +170,10 @@ def normalize_config(cfg):
             cfg.processor_config or cfg.base_model_config or cfg.base_model
         )
-        model_config = model_config.text_config
+        if hasattr(model_config, "text_config"):
+            model_config = model_config.text_config
+        elif hasattr(model_config, "get_text_config"):
+            model_config = model_config.get_text_config()
```
I think this is set twice. Here and in `load_model_config()->check_model_config()` above

NanoCode012 commented on src/axolotl/processing_strategies/__init__.py:
```diff
@@ -0,0 +1,210 @@
+        for i, message in enumerate(messages):
+            if message["role"] == "user":
+                for j, content in enumerate(message["content"]):
+                    if "type" in content and content["type"] == "text":
+                        messages[i]["content"][j] = {
+                            "type": "text",
+                            "content": content["text"],
+                        }
```
Do we not need to add handling for images as well?

Edit: Why do we need this processing in particular?
